go/libraries/doltcore/doltdb/table_editor.go

```diff
@@ -167,181 +224,169 @@ func (te *TableEditor) Update(ctx context.Context, dOldRow row.Row, dNewRow row.
 		}
 
 		// If the old value of the primary key we just updated was previously inserted, then we need to remove it now.
-		if _, ok := te.insertedKeys[oldHash]; ok {
-			delete(te.insertedKeys, oldHash)
-			te.ed.Remove(dOldKeyVal)
+		if _, ok := te.tea.insertedKeys[oldHash]; ok {
+			delete(te.tea.insertedKeys, oldHash)
+			te.tea.ed.AddEdit(dOldKeyVal, nil)
+			te.tea.opCount++
 		}
 
-		te.addedKeys[newHash] = dNewKeyVal
-		te.removedKeys[oldHash] = dOldKeyVal
-		te.affectedKeys[oldHash] = dOldKeyVal
+		te.tea.addedKeys[newHash] = dNewKeyVal
+		te.tea.removedKeys[oldHash] = dOldKeyVal
+		te.tea.affectedKeys[oldHash] = dOldKeyVal
 	}
 
-	te.affectedKeys[newHash] = dNewKeyVal
+	te.tea.affectedKeys[newHash] = dNewKeyVal
 
-	te.ed.Set(dNewKeyVal, dNewRow.NomsMapValue(te.tSch))
-	te.updated = true
+	te.tea.ed.AddEdit(dNewKeyVal, dNewRow.NomsMapValue(te.tSch))
+	te.tea.opCount++
+	te.autoFlush()
 	return nil
 }
 
-// Flush finalizes all of the changes and returns the updated Table.
-func (te *TableEditor) Flush(ctx context.Context) (*Table, error) {
+// Flush finalizes all of the changes made so far.
+func (te *TableEditor) Flush() {
 	te.flushMutex.Lock()
 	defer te.flushMutex.Unlock()
 
-	if !te.updated {
-		return te.t, nil
+	if te.tea.opCount > 0 {
+		te.aq.Append(te.tea)
+		te.tea = newTableEditAcc(te.nbf)
+	}
+}
+
+// Table returns a Table based on the edits given, if any. If Flush() was not called prior, it will be called here.
+func (te *TableEditor) Table() (*Table, error) {
+	te.Flush()
+	err := te.aq.WaitForEmpty()
+	return te.t, err
+}
+
+// autoFlush is called from the end of write calls, so we must Unlock all held locks before calling Flush. We re-lock
```

Comment:
It seems like you can probably remove the flush locks from the various tableEditor methods and only deal with them in two places: 1) the AddEdit method (i suggested this method above) and 2) the Flush method. At this point the flushLock is just guarding the access to the action queue and assignment to the te.tea field, which both are only handled in the Flush method. You don't need a lock in the autoFlush method at all, because te.tea.opCount is only ever incremented or set back to zero.